mmengine/dist/utils.py

```diff
@@ -537,3 +544,137 @@ def cast_data_device(
+            Defaults to 0.
+        master_addr (str, optional): The FQDN of the host that is running
+            worker with rank 0; used to initialize the Torch Distributed
+            backend. Defaults to '127.0.0.1'.
+        master_port (str, optional): The port on the ``master_addr`` that can
+            be used to host the C10d TCP store. Defaults to 'auto'.
+        args (tuple, optional): Arguments passed to main_func. Defaults to ().
+    """
+    if not isinstance(num_proc_per_machine, list):
+        num_proc_per_machine = [num_proc_per_machine] * num_machines
+    world_size = sum(num_proc_per_machine)
+
+    if master_port == 'auto':
+        assert len(num_proc_per_machine) == 1, (
+            'Automatically inferring free port on multiple machines will lead'
+            'to potential error.')
+        master_port = str(_find_free_port())
+
+    if world_size > 1:
+        # https://github.com/pytorch/pytorch/pull/14391
+
+        mp.start_processes(
+            _distributed_worker,
+            nprocs=num_proc_per_machine[machine_rank],
+            args=(
+                main_func,
+                world_size,
+                num_proc_per_machine,
+                machine_rank,
+                master_addr,
+                master_port,
+                args,
+            ),
+            daemon=False,
+        )
+    else:
+        main_func(*args)
+
+
+def _distributed_worker(
+    local_rank: int,
+    main_func: Callable,
+    world_size: int,
+    num_proc_per_machine: list,
+    machine_rank: int,
+    master_addr: str,
+    master_port: str,
+    args,
+) -> None:
+    """Run the task after initializing the environment.
+
+    This function will be launched by ``mp.start_processes`` and ``local_rank``
+    will be filled automatically.
+
+    Part of environment variable defined in `pytorch official docs <https://pytorch.org/docs/stable/elastic/run.html#environment-variables>`_
+    will be initialized.
+
+    Warning:
+        :func:`init_dist` must be called in the ``main_func``
+
+    Args:
+        local_rank (int): Local rank of the current machine.
+        main_func (Callable): Function to be executed.
+        world_size (int): The number of all processes launched on all machines.
+        num_proc_per_machine (list): Number of launched processes of all
+            machines.
+        machine_rank (int): The rank of current machine.
+        master_addr (str): The FQDN of the host that is running
+            worker with rank 0; used to initialize the Torch Distributed
+            backend.
+        master_port (str): The port on the ``master_addr`` that can
+            be used to host the C10d TCP store.
+        args (tuple, optional): Arguments passde to main_func. Defaults to ().
+    """  # noqa: E501
+    has_gpu = torch.cuda.is_available()
+    if has_gpu:
+        assert num_proc_per_machine[machine_rank] <= torch.cuda.device_count()
```

Comment:
We need to consider other chips like Ascend or MLU.

Reply:
I'm not sure whether chips like MLU or NPU can enable ddp-training in this way, and whether their specialized PyTorch has a general interface to get this device information. I think this function could be extended in the future to support other chips.